File: api-gateway/src/config/logger.js
const winston = require('winston');

const { combine, timestamp, printf, colorize, errors } = winston.format;

const logFormat = printf(({ level, message, timestamp, stack}) => {
    return `[${timestamp}] ${level.toUpperCase()}: ${stack || message}`;
});

const logger = winston.createLogger({
    level: process.env.LOG_LEVEL || 'info',
    format: combine(
        timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
        errors({ stack: true}),
        logFormat
    ),
    transports: [
        new winston.transports.Console({
            format: combine(
                colorize({ all: true}),
                timestamp({ format: 'YYYY-MM-DD HH:mm:ss' }),
                errors({ stack: true}),
                logFormat
            ),
        }),
        new winston.transports.File({ filename: 'logs/combined.log' }),
        new winston.transports.File({ filename: 'logs/error.log', level: 'error' }),
    ]
});

module.exports = logger;

File: order-service/app/config/logger.py
import logging
import os
from pathlib import Path

# Create logs directory if it does not exist
Path("logs").mkdir(exist_ok=True)


# Get log level from the environment variable - default to INFO
LOG_LEVEL = os.getenv("LOG_LEVEL", "INFO").upper()

# Create the logger
logger = logging.getLogger("order-service")
logger.setLevel(LOG_LEVEL)

# -- Formatter ------------------------
# Same format as our Node.js services for consistency
formatter = logging.Formatter(
    "[%(asctime)s %(levelname)s: %(message)s]",
    datefmt="%Y-%m-%d %H:%M:%S"
)

# -- Console Handler ------------------
# Prints logs to terminal
console_handler = logging.StreamHandler()
console_handler.setFormatter(formatter)
logger.addHandler(console_handler)

# -- File Handlers --------------------
# All logs -> combined.log
file_handler = logging.FileHandler("logs/combined.log")
file_handler.setFormatter(formatter)
logger.addHandler(file_handler)

# Only errors -> error.log
error_handler = logging.FileHandler("logs/error.log")
error_handler.setLevel(logging.ERROR)
error_handler.setFormatter(formatter)
logger.addHandler(error_handler)